### ISSUE TITLE
Reactor overwrites `user` argument when calling runner or wheel module

### ISSUE THREAD

**Issue by vutny** [saltstack/salt#30604]
Hello there,

I want to pass the `user` parameter to my custom runner in reactor SLS:

```
my-runner: 
  runner.my.custom_runner: 
    - user: me
```

It seems that the parameter's value `me` is always going to be replaced with `Reactor`.

The code responsible for this is here: https://github.com/saltstack/salt/blob/develop/salt/utils/reactor.py#L264

```
    def run(self, low):
        '''
        Execute the specified function in the specified state by passing the
        LowData
        '''
        l_fun = getattr(self, low['state'])
        try:
            f_call = salt.utils.format_call(l_fun, low)
            kwargs = f_call.get('kwargs', {}) 

            # TODO: Setting the user doesn't seem to work for actual remote publishes
            if low['state'] in ('runner', 'wheel'):
                # TODO: pick one...
                kwargs['__user__'] = self.event_user
                kwargs['user'] = self.event_user

            l_fun(*f_call.get('args', ()), **kwargs)
        except Exception:
            log.error(
                    'Failed to execute {0}: {1}\n'.format(low['state'], l_fun),
                    exc_info=True
                    )
```

I don't know why we need to supply `event_user` property to the function call.
Of course because this is my runner, I could replace `user` argument with something similar, but this kind of reactor's behavior is totally not clear and it may potentially affect some other runner or wheel modules.

I suggest to remove this line:
```
kwargs['user'] = self.event_user
```

Looks like this wouldn't affect anything else. I can open a PR if this is acceptable.

**Comment by jfindlay** [comment]
@vutny, thanks for your work on this.  @jacksontj, do you have any comments here?

**Comment by vutny** [comment]
@jfindlay @jacksontj There are the only places in the whole Salt code where `event_user` property is ever used:

```python
./salt/utils/reactor.py:217:            if data['data'].get('user') == self.wrap.event_user:
./salt/utils/reactor.py:252:    event_user = 'Reactor'
./salt/utils/reactor.py:277:                kwargs['__user__'] = self.event_user
./salt/utils/reactor.py:278:                kwargs['user'] = self.event_user
```

If I got it right, `Reactor` user were introduced to avoid processing the event fired by reactor itself (line 217).
Anyhow, after removing `kwargs['user'] = self.event_user` line I still see the user in the events:

```python
salt/run/20160126100057030906/ret	{
    "_stamp": "2016-01-26T10:01:02.538633", 
    "fun": "runner.my.custom_runner", 
    "jid": "20160126100057030906", 
    "return": null, 
    "success": true, 
    "user": "Reactor"
}
```

That's correct and expected. See the `SyncClientMixin` class:

```python
./salt/client/mixins.py-233-    def low(self, fun, low):
./salt/client/mixins.py-234-        '''
./salt/client/mixins.py-235-        Execute a function from low data
./salt/client/mixins.py-236-        Low data includes:
./salt/client/mixins.py-237-            required:
./salt/client/mixins.py-238-                - fun: the name of the function to run
./salt/client/mixins.py-239-            optional:
./salt/client/mixins.py-240-                - args: a list of args to pass to fun
./salt/client/mixins.py-241-                - kwargs: kwargs for fun
./salt/client/mixins.py:242:                - __user__: user who is running the command
./salt/client/mixins.py-243-                - __jid__: jid to run under
./salt/client/mixins.py-244-                - __tag__: tag to run under
./salt/client/mixins.py-245-        '''
./salt/client/mixins.py-246-        # fire the mminion loading (if not already done) here
./salt/client/mixins.py-247-        # this is not to clutter the output with the module loading
./salt/client/mixins.py-248-        # if we have a high debug level.
./salt/client/mixins.py-249-        self.mminion  # pylint: disable=W0104
./salt/client/mixins.py-250-        jid = low.get('__jid__', salt.utils.jid.gen_jid())
./salt/client/mixins.py-251-        tag = low.get('__tag__', tagify(jid, prefix=self.tag_prefix))
./salt/client/mixins.py-252-
./salt/client/mixins.py-253-        data = {'fun': '{0}.{1}'.format(self.client, fun),
./salt/client/mixins.py-254-                'jid': jid,
./salt/client/mixins.py:255:                'user': low.get('__user__', 'UNKNOWN'),
./salt/client/mixins.py-256-                }
```

The low data argument `__user__` value will be passed to the event data as `user`.
I think this proves that removing `user` argument overwriting from reactor actually doesn't break anything and will help to avoid future confusions.

Thank you for reading, I would like to see your comments.

**Comment by jacksontj** [comment]
As you mentioned, this feature was added to avoid loops. From your comment it sounds like you removed some of the code, but not all. If you remove those 3 lines-- and have a reactor that runs on every event you will infinitely recurse. Now, if your SLSs wouldn't run on reactor fired events-- then you'd be fine, but if your glob caught it then you'd have issues (which is what the fix was for). If we really want/need to preserve the `user` field, then we could look into some other mechanism to *not* fire events from the reactor.

**Comment by vutny** [comment]
@jacksontj Thank you very much for your explanation. Now I have clearer understanding of how this `event --> reactor --> function --> event` chain works internally. Still I can't get the purpose of passing `user` kwarg to the function's low data (while `__user__` is obviously correct). 

Let me describe the flow as I see it, please correct me if I'm wrong.
# TL;DR
`kwargs['user'] = self.event_user` line is aimless.

Even if I change it to something like:
```
kwargs['user'] = 'Wrong User'
```
I still see the correct user `Reactor` in the event data, which is the guarantee for escaping the event loop. Please read the details below.

# Reactor event flow:
1. Event fired with some tag and other data.
2. I have a Reactor SLS set in the Salt Mater config matching the event tag
3. Salt Reactor "reads" my runner (or wheel) function with arguments from SLS in the `run` method of `ReactWrap` class (`salt/utils/reactor.py`): 

```python
246 class ReactWrap(object):
247     '''
248     Create a wrapper that executes low data for the reaction system
249     '''
250     # class-wide cache of clients
251     client_cache = None
252     event_user = 'Reactor'
253 
...
264     def run(self, low):
265         '''
266         Execute the specified function in the specified state by passing the
267         LowData
268         '''
269         l_fun = getattr(self, low['state'])
270         try:
271             f_call = salt.utils.format_call(l_fun, low)
272             kwargs = f_call.get('kwargs', {})
```
 
4.After that Reactor checks the low data to find a call to `runner` or `wheel` modules.

```python
274             # TODO: Setting the user doesn't seem to work for actual remote publishes
275             if low['state'] in ('runner', 'wheel'):
276                 # TODO: pick one...
277                 kwargs['__user__'] = self.event_user
278                 kwargs['user'] = self.event_user
```

And here it also forcibly overrides two arguments in the future function call.
5. Now the actual function is called:
```python
280             l_fun(*f_call.get('args', ()), **kwargs)
281         except Exception:
282             log.error(
283                     'Failed to execute {0}: {1}\n'.format(low['state'], l_fun),
284                     exc_info=True
285                     )
```
And new event about the function was called is going to be fired.
6. How the event data will be populated? Take a look at `salt/client/mixins.py`:
```python
106 class SyncClientMixin(object):
107     '''
108     A mixin for *Client interfaces to abstract common function execution
109     '''
...
233     def low(self, fun, low):
234         '''
235         Execute a function from low data
236         Low data includes:
237             required:
238                 - fun: the name of the function to run
239             optional:
240                 - args: a list of args to pass to fun
241                 - kwargs: kwargs for fun
242                 - __user__: user who is running the command
243                 - __jid__: jid to run under
244                 - __tag__: tag to run under
245         '''
246         # fire the mminion loading (if not already done) here
247         # this is not to clutter the output with the module loading
248         # if we have a high debug level.
249         self.mminion  # pylint: disable=W0104
250         jid = low.get('__jid__', salt.utils.jid.gen_jid())
251         tag = low.get('__tag__', tagify(jid, prefix=self.tag_prefix))
252 
253         data = {'fun': '{0}.{1}'.format(self.client, fun),
254                 'jid': jid,
255                 'user': low.get('__user__', 'UNKNOWN'),
256                 }
257 
```
See it? `user` value set the `data` dict is taken from `__user__` argument in the low data (provided by Reactor). On other hand, `user` argument in low data isn't going anywhere.
7. Finally, Salt fires the event.
```python
258         event = salt.utils.event.get_event(
259                 'master',
260                 self.opts['sock_dir'],
261                 self.opts['transport'],
262                 opts=self.opts,
263                 listen=False)
264 
265         namespaced_event = salt.utils.event.NamespacedEvent(
266             event,
267             tag,
268             print_func=self.print_async_event
269                        if hasattr(self, 'print_async_event')
270                        else None
271         )
272         # TODO: document these, and test that they exist
273         # TODO: Other things to inject??
274         func_globals = {'__jid__': jid,
275                         '__user__': data['user'],
276                         '__tag__': tag,
277                         # weak ref to avoid the Exception in interpreter
278                         # teardown of event
279                         '__jid_event__': weakref.proxy(namespaced_event),
280                         }
281 
282         func_globals['__jid_event__'].fire_event(data, 'new')
```
`user` value is taken from the event data here, which is `__user__` in function's low data assigned above, not the `user`. I hope I made it comprehensible :-)
8. OK, what if I got a Reactor configured to watch the event generated by function called by Reactor? This is where event loop is possible as you mentioned. Let's get back to `salt/utils/reactor.py`:
```python
 24 class Reactor(salt.utils.process.SignalHandlingMultiprocessingProcess, salt.state.Compiler):
 25     '''
 26     Read in the reactor configuration variable and compare it to events
 27     processed on the master.
 28     The reactor has the capability to execute pre-programmed executions
 29     as reactions to events
 30     '''
...
200     def run(self):
201         '''
202         Enter into the server loop
203         '''
204         salt.utils.appendproctitle(self.__class__.__name__)
205 
206         # instantiate some classes inside our new process
207         self.event = salt.utils.event.get_event(
208                 'master',
209                 self.opts['sock_dir'],
210                 self.opts['transport'],
211                 opts=self.opts,
212                 listen=True)
213         self.wrap = ReactWrap(self.opts)
214 
215         for data in self.event.iter_events(full=True):
216             # skip all events fired by ourselves
217             if data['data'].get('user') == self.wrap.event_user:
218                 continue
```
This is the main point in last two lines. Salt will skip the events whose data dict contains `user` value matches `Reactor`. As you remember, `user` value is taken exclusively from `__user__` low data argument.

As a conclusion, the line https://github.com/saltstack/salt/blob/6ce6b21990a59e37dbf7d60c11829abbdf7868cc/salt/utils/reactor.py#L278
`kwargs['user'] = self.event_user` affects only function's low data, not the event data.
And in my opinion it's very confusing and purposeless.

I keep suggesting to remove the line :-) Thanks for your time and understanding.

**Comment by cachedout** [comment]
Closed via #30815
